workspace-manager-clienttests/src/main/java/scripts/uploadscripts/SummariesToBigQuery.java

```diff
@@ -0,0 +1,215 @@
+        }
+        projectId = parameters.get(0);
+        datasetName = parameters.get(1);
+    }
+
+    private static String testRunTableName = "testRun";
+    private static String testScriptResultsTableName = "testScriptResults";
+    private static String measurementCollectionTableName = "measurementCollection";
+
+    /**
+     * Upload the test results saved to the given directory. Results may include Test Runner
+     * client-side output and any relevant measurements collected.
+     */
+    public void uploadResults(
+            Path outputDirectory, ServiceAccountSpecification uploaderServiceAccount) throws Exception {
+        // get a BigQuery client object
+        logger.debug("BigQuery project_id:dataset_name: {}:{}", projectId, datasetName);
+        BigQuery bigQueryClient =
+                BigQueryUtils.getClientForServiceAccount(uploaderServiceAccount, projectId);
+        // read in TestConfiguration TestRunSummary, and array of
+        // MeasurementCollectionScript.MeasurementResultSummary objects
+        renderedTestConfiguration = TestRunner.getRenderedTestConfiguration(outputDirectory);
+        testRunSummary = TestRunner.getTestRunSummary(outputDirectory);
+        measurementCollectionSummaries =
+                MeasurementCollector.getMeasurementCollectionSummaries(outputDirectory);
+
+        // insert a single row into testRun
+        if (BigQueryUtils.checkRowExists(
+                bigQueryClient, projectId, datasetName, testRunTableName, "id", testRunSummary.id)) {
+            logger.info(
+                    "A row with this id already exists in the "
+                            + testRunTableName
+                            + " table. Inserting a duplicate.");
+        }
+        TableId tableId = TableId.of(datasetName, testRunTableName);
+        InsertAllRequest insertRequest =
+                InsertAllRequest.newBuilder(tableId).addRow(buildTestRunRow(outputDirectory)).build();
+        BigQueryUtils.insertAllIntoBigQuery(bigQueryClient, insertRequest);
+
+        // insert into testScriptResults
+        tableId = TableId.of(datasetName, testScriptResultsTableName);
+        InsertAllRequest.Builder insertRequestBuilder = InsertAllRequest.newBuilder(tableId);
+        for (int ctr = 0; ctr < renderedTestConfiguration.testScripts.size(); ctr++) {
+            TestScriptSpecification testScriptSpecification =
+                    renderedTestConfiguration.testScripts.get(ctr);
+            TestScriptResult.TestScriptResultSummary testScriptResult =
+                    testRunSummary.testScriptResultSummaries.get(ctr);
+            insertRequestBuilder.addRow(
+                    buildTestScriptResultsRow(testScriptSpecification, testScriptResult));
+        }
+        BigQueryUtils.insertAllIntoBigQuery(bigQueryClient, insertRequestBuilder.build());
+
+        // insert into measurementCollection
+        if (measurementCollectionSummaries == null) {
```

Comment:
nit: it would be better to make sure this is never null and initialize to an empty list. Note that you can also have no summaries if the collection is empty.

Reply:
Added your comment to enhancement [doc](https://docs.google.com/document/d/1gYqlJTpC4C4S1K0pYikeNJx-HWNZAZJYybinmuqcSTI/edit#). This code will likely end up in the Test Runner library itself once all the enhancement considerations are taken into account.